risc0/zkvm/src/host/server/prove/mod.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Run the zkVM guest and prove its results.
+//!
+//! # Usage
+//! The primary use of this module is to provably run a zkVM guest by use of a
+//! [Session]. See the [Session] documentation for more detailed usage
+//! information.
+//!
+//! ```rust
+//! use risc0_zkvm::{default_prover, ExecutorEnv};
+//! use risc0_zkvm_methods::FIB_ELF;
+//!
+//! # #[cfg(not(feature = "cuda"))]
+//! # {
+//! let env = ExecutorEnv::builder().add_input(&[20]).build().unwrap();
+//! let receipt = default_prover().prove_elf(env, FIB_ELF).unwrap();
+//! # }
+//! ```
+
+mod dev_mode;
+mod exec;
+pub(crate) mod loader;
+mod plonk;
+mod prover_impl;
+#[cfg(test)]
+mod tests;
+
+use std::rc::Rc;
+
+use anyhow::Result;
+use cfg_if::cfg_if;
+use risc0_binfmt::{MemoryImage, Program};
+use risc0_circuit_rv32im::CircuitImpl;
+use risc0_core::field::{
+    baby_bear::{BabyBear, Elem, ExtElem},
+    Elem as _,
+};
+use risc0_zkp::{
+    adapter::CircuitInfo,
+    core::digest::DIGEST_WORDS,
+    hal::{EvalCheck, Hal},
+};
+use risc0_zkvm_platform::{memory::MEM_SIZE, PAGE_SIZE, WORD_SIZE};
+
+use self::{dev_mode::DevModeProver, prover_impl::ProverImpl};
+use crate::{
+    is_dev_mode, Executor, ExecutorEnv, ProverOpts, Receipt, Segment, SegmentReceipt, Session,
+    VerifierContext,
+};
+
+/// A DynProverImpl can execute a given [MemoryImage] and produce a [Receipt]
+/// that can be used to verify correct computation.
+pub trait DynProverImpl {
```

Comment:
It took me a minute to understand that `Prover` is the client and `DynProverImpl` is the server. (Also it's slightly confusing that a type with the name `Impl` is a `trait` instead of a `struct`). Maybe this should be something like `ProverServer`?

Reply:
Naming is hard :) Maybe `ProverServer`

Reply:
I like ProverSever!